tutorial/decorators.md

```diff
@@ -0,0 +1,243 @@
+
+  contains
+
+  !........................................
+  subroutine fun1()
+
+    implicit none
+
+    integer(i64) :: array_in_stack(0:2_i64)
+    integer(i64), allocatable :: array_in_heap(:)
+
+    !/////////////////////////
+    !array stored in the stack
+    !////////////////////////
+    array_in_stack = [1_i64, 2_i64, 3_i64]
+    !////////////////////////
+    !array stored in the heap
+    !////////////////////////
+    allocate(array_in_heap(0:2_i64))
+    array_in_heap = [1_i64, 2_i64, 3_i64]
+
+  end subroutine fun1
+  !........................................
+
+end module boo
+```
+
+## Allow negative index
+
+This one indicates that all arrays mentioned as argements (of the decorator) can be accessed with negative indexes. unlike fortran you can't use negative indexes in arrays in C, But with Pyccel you can work comfotabally with them in the Python code, For the generated C code Pyccel will take care of that.
+
+An example shows how pyccel handles negative indexes beween Python and C:
+
+```python
+from pyccel.decorators import allow_negative_index
+
+@allow_negative_index('a')
+def fun1():
+    a = [1,2,3,4,5,6]
+    print(a[-1])
+
+    b = [1,2,3,4,5,6]
+    print(b[-1])
+```
+
+This is the generated C code:
+
+```C
+
+#include "boo.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include "ndarrays.h"
+
+
+/*........................................*/
+void fun1(void)
+{
+    t_ndarray a;
+    t_ndarray b;
+    a = array_create(1, (int64_t[]){6}, nd_int64);
+    int64_t array_dummy_0001[] = {1, 2, 3, 4, 5, 6};
+    memcpy(a.nd_int64, array_dummy_0001, a.buffer_size);
+    printf("%ld\n", GET_ELEMENT(a, nd_int64, 5));
+    b = array_create(1, (int64_t[]){6}, nd_int64);
+    int64_t array_dummy_0002[] = {1, 2, 3, 4, 5, 6};
+    memcpy(b.nd_int64, array_dummy_0002, b.buffer_size);
+    printf("%ld\n", GET_ELEMENT(b, nd_int64, 5));
+}
+/*........................................*/
+```
+
+And here is the equivalent Fortran code:
+
+```fortran
+module boo
+
+
+  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+  implicit none
+
+  contains
+
+  !........................................
+  subroutine fun1()
+
+    implicit none
+
+    integer(i64), allocatable :: a(:)
+    integer(i64), allocatable :: b(:)
+
+    allocate(a(0:5_i64))
+    a = [1_i64, 2_i64, 3_i64, 4_i64, 5_i64, 6_i64]
+    print *, a(size(a, kind=i64) - 1_i64)
+    allocate(b(0:5_i64))
+    b = [1_i64, 2_i64, 3_i64, 4_i64, 5_i64, 6_i64]
+    print *, b(size(b, kind=i64) - 1_i64)
+
+  end subroutine fun1
+  !........................................
+
+end module boo
+```
+
+## Elemental and pure decorators
+
+This decorator indicates that the function bellow is an elemental one, An elemental function is a function with a single scalar operator and a scalar return value which can also be called on an array. When it is called on an array it returns the result of the function called elementwise on the array.
```

Comment:
"This decorator" `pure` or `elemental`?

Reply:
This description is definitely for `elemental` functions. `pure` has to do with the absence of side effects, so a "pure" function is the "standard" function in purely functional languages.